File: graph/evaluate_division.py
from typing import List
from collections import defaultdict, deque


class Solution:
    # dfs
    def calcEquation1(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        edges = defaultdict(dict)
        for (pt1, pt2), val in zip(equations, values):
            edges[pt1][pt2] = val
            edges[pt2][pt1] = 1/val

        visited = set()
        result = []

        def dfs(cur, dest, totalVal):
            if cur == dest:
                return totalVal
            for nei, val in edges[cur].items():
                if nei not in visited:
                    visited.add(nei)
                    res = dfs(nei, dest, totalVal * val)
                    if res != -1:
                        return res
                    visited.remove(nei)
            return -1

        for src, dest in queries:
            result.append(dfs(src, dest, 1) if src in edges else -1)
            visited = set()

        return result

    # bfs
    def calcEquation2(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        edges = {}
        for (pt1, pt2), val in zip(equations, values):
            if pt1 not in edges:
                edges[pt1] = {pt2 : val}
            else:
                edges[pt1][pt2] = val
            if pt2 not in edges:
                edges[pt2] = {pt1 : 1/val}
            else:
                edges[pt2][pt1] = 1/val
        result = []

        for src, dest in queries:
            q = deque()
            visitSet = set()
            totalVal = -1
            if src in edges:
                q.append((src, 1))
                visitSet.add(src)
            while q:
                cur, val = q.popleft()
                if cur == dest:
                    totalVal = val
                    break
                for nei, neiVal in edges[cur].items():
                    if nei not in visitSet:
                        visitSet.add(nei)
                        q.append((nei, val * neiVal))
            result.append(totalVal)

        return result

File: graph/evaluate_division.java
public class evaluate_division {
    
}

class Solution {
    public double[] calcEquation1(List<List<String>> equations, double[] values, List<List<String>> queries) {
        Map<String, Map<String, Double>> edges = new HashMap<>();
        for(int i = 0; i < equations.size(); i++){
            String pt1 = equations.get(i).get(0);
            String pt2 = equations.get(i).get(1);
            double val = values[i];

            edges.putIfAbsent(pt1, new HashMap());
            edges.putIfAbsent(pt2, new HashMap());

            edges.get(pt1).put(pt2, val);
            edges.get(pt2).put(pt1, 1/val);
        }

        double[] result = new double[queries.size()];
        
        for(int i = 0; i < queries.size(); i++){
            String src = queries.get(i).get(0);
            String dest = queries.get(i).get(1);
            result[i] = dfs(src, dest, edges, new HashSet<>());
        }

        return result;
    }

    public double dfs(String cur, String dest, Map<String, Map<String, Double>> edges, Set<String> visited){
        if (!edges.containsKey(cur) || !edges.containsKey(dest))
            return -1.0;
        
        if (cur.equals(dest))
            return 1.0;
        
        visited.add(cur);
        
        Map<String, Double> neighbors = edges.get(cur);
        for (Map.Entry<String, Double> entry : neighbors.entrySet()) {
            String neighbor = entry.getKey();
            double val = entry.getValue();
            
            if (!visited.contains(neighbor)) {
                double result = dfs(neighbor, dest, edges, visited);
                if (result != -1.0)
                    return val * result;
            }
        }
        
        visited.remove(cur);
        
        return -1.0;
    }

    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries){
        Map<String, Map<String, Double>> edges = new HashMap<>();
        for(int i = 0; i < equations.size(); i++){
            String pt1 = equations.get(i).get(0);
            String pt2 = equations.get(i).get(1);
            double val = values[i];

            edges.putIfAbsent(pt1, new HashMap());
            edges.putIfAbsent(pt2, new HashMap());

            edges.get(pt1).put(pt2, val);
            edges.get(pt2).put(pt1, 1/val);
        }

        double[] result = new double[queries.size()];
    
        for(int i = 0; i < queries.size(); i++){
            String src = queries.get(i).get(0);
            String dest = queries.get(i).get(1);
            Queue<Pair<String, Double>> queue = new LinkedList<>();
            Set<String> visited = new HashSet<>();
            double totalVal = -1.0;
            if(edges.containsKey(src)){
                queue.offer(new Pair<>(src, 1.0));
                visited.add(src);
            }
            while(!queue.isEmpty()){
                Pair<String, Double> pair = queue.poll();
                String cur = pair.getKey();
                Double val = pair.getValue();
                if(cur.equals(dest)){
                    totalVal = val;
                    break;
                }
                Map<String, Double> neighbors = edges.get(cur);
                for (Map.Entry<String, Double> entry : neighbors.entrySet()){
                    String neighbor = entry.getKey();
                    double neighborVal = entry.getValue();
                    if(!visited.contains(neighbor)){
                        visited.add(neighbor);
                        queue.offer(new Pair<>(neighbor, val * neighborVal));
                    }
                }
            }
            result[i] = totalVal;
        }

        return result;

    }

}